src/main/java/org/passay/RuleResultMetadata.java

```diff
@@ -13,77 +13,127 @@
+    Length,
+
+    /** lowercase characters. */
+    LowerCase,
+
+    /** uppercase characters. */
+    UpperCase,
+
+    /** digit characters. */
+    Digit,
+
+    /** special characters. */
+    Special,
+
+    /** whitespace characters. */
+    Whitespace,
+
+    /** allowed characters. */
+    Allowed,
+
+    /** illegal characters. */
+    Illegal;
+
+
+    /**
+     * Returns whether a count category exists with the supplied name.
+     *
+     * @param  name  to check.
+     *
+     * @return  whether the supplied name exists.
+     */
+    public static boolean exists(final String name)
+    {
+      for (CountCategory cc : CountCategory.values()) {
+        if (cc.name().equals(name)) {
+          return true;
+        }
+      }
+      return false;
+    }
+  }
+
+  /** Character count metadata. */
+  protected final Map<CountCategory, Integer> counts = new HashMap<>();
+
+
+  /**
+   * Creates a new rule result metadata.
+   */
+  public RuleResultMetadata() {}
 
 
   /**
    * Creates a new rule result metadata.
    *
-   * @param  data  metadata.
+   * @param  category  count category.
+   * @param  value  count value.
    */
-  public RuleResultMetadata(final Map<String, Object> data)
+  public RuleResultMetadata(final CountCategory category, final int value)
   {
-    if (data == null) {
-      metadata = new LinkedHashMap<>();
-    } else {
-      metadata = new LinkedHashMap<>(data);
-    }
+    putCount(category, value);
   }
 
 
   /**
-   * Puts the supplied key and value into this metadata object.
+   * Returns the count for the supplied category.
+   *
+   * @param  category  of the count.
    *
-   * @param  key  of the metadata.
-   * @param  value  of the metadata.
+   * @return  character count.
    */
-  public void put(final String key, final Object value)
+  public int getCount(final CountCategory category)
   {
-    metadata.put(key, value);
+    return counts.get(category);
   }
 
 
   /**
-   * Puts all the supplied metadata into this metadata object.
+   * Returns an unmodifiable map of all count metadata.
    *
-   * @param  data  to add.
+   * @return  count metadata.
    */
-  public void putAll(final Map<String, Object> data)
+  public Map<CountCategory, Integer> getCounts()
   {
-    metadata.putAll(data);
+    return Collections.unmodifiableMap(counts);
   }
 
 
   /**
-   * Returns the metadata value for the supplied key.
+   * Adds a count to the metadata.
    *
-   * @param  <T>  type of value to return
-   * @param  key  to the metadata to retrieve.
-   * @param  type  to cast the metadata value to.
-   *
-   * @return  metadata value.
+   * @param  category  of the count.
+   * @param  value  of the count.
    */
-  public <T> T get(final String key, final Class<T> type)
+  public void putCount(final CountCategory category, final int value)
   {
-    return type.cast(metadata.get(key));
+    if (value < 0) {
+      throw new IllegalArgumentException("Count value must be greater than zero");
```

Comment:
Zero count is allowed, so error message is misleading. "Count must be non-negative" seems sufficiently clear and correct. Also, might add a note about allowed range to the `value` parameter: `@param  value  non-negative character count`.

Reply:
Fixed.